include/QuantumComputation.hpp

```diff
@@ -106,55 +170,95 @@ namespace qc {
+        Permutation initialLayout{};
+        Permutation outputPermutation{};
+
+        std::vector<bool> ancillary{};
+        std::vector<bool> garbage{};
+
+        [[nodiscard]] std::size_t getNindividualOps() const;
+
+        [[nodiscard]] std::string                         getQubitRegister(dd::Qubit physical_qubit_index) const;
+        [[nodiscard]] std::string                         getClassicalRegister(std::size_t classical_index) const;
+        static dd::Qubit                                  getHighestLogicalQubitIndex(const Permutation& map);
+        [[nodiscard]] dd::Qubit                           getHighestLogicalQubitIndex() const { return getHighestLogicalQubitIndex(initialLayout); };
+        [[nodiscard]] std::pair<std::string, dd::Qubit>   getQubitRegisterAndIndex(dd::Qubit physical_qubit_index) const;
+        [[nodiscard]] std::pair<std::string, std::size_t> getClassicalRegisterAndIndex(std::size_t classical_index) const;
+
+        [[nodiscard]] dd::Qubit   getIndexFromQubitRegister(const std::pair<std::string, dd::Qubit>& qubit) const;
+        [[nodiscard]] std::size_t getIndexFromClassicalRegister(const std::pair<std::string, std::size_t>& clbit) const;
+        [[nodiscard]] bool        isIdleQubit(dd::Qubit physical_qubit) const;
+        [[nodiscard]] bool        isLastOperationOnQubit(const decltype(ops.cbegin())& opIt, const decltype(ops.cend())& end) const;
+        [[nodiscard]] bool        physicalQubitIsAncillary(dd::Qubit physical_qubit_index) const;
+        [[nodiscard]] bool        logicalQubitIsAncillary(dd::Qubit logical_qubit_index) const { return ancillary[logical_qubit_index]; }
+        void                      setLogicalQubitAncillary(dd::Qubit logical_qubit_index) { ancillary[logical_qubit_index] = true; }
+        [[nodiscard]] bool        logicalQubitIsGarbage(dd::Qubit logical_qubit_index) const { return garbage[logical_qubit_index]; }
+        void                      setLogicalQubitGarbage(dd::Qubit logical_qubit_index) { garbage[logical_qubit_index] = true; }
+        MatrixDD                  createInitialMatrix(std::unique_ptr<dd::Package>& dd) const; // creates identity matrix, which is reduced with respect to the ancillary qubits
 
         /// strip away qubits with no operations applied to them and which do not pop up in the output permutation
         /// \param force if true, also strip away idle qubits occurring in the output permutation
         void stripIdleQubits(bool force = false, bool reduceIOpermutations = true);
         // apply swaps 'on' DD in order to change 'from' to 'to'
         // where |from| >= |to|
-        static void changePermutation(dd::Edge& on, qc::permutationMap& from, const qc::permutationMap& to, std::array<short, qc::MAX_QUBITS>& line, std::unique_ptr<dd::Package>& dd, bool regular = true);
+        template<class DD>
```

Comment:
The name `DD` is a bit unlucky, as this this refers to vector edges or matrix edges. How about `template <class EdgeType>`?

Reply:
What about `DDType`? I feel that the concept of matrix and vector decision diagrams is better in the context of the QFR than "edges" (which are fine in the DD package itself).

Reply:
Works for me.